### PR TITLE
MapSnapshot attribution

### PR DESCRIPTION
**Pull request by tobrun**
ezgif com-video-to-gif

### CODE REVIEW

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/attribution/AttributionMeasure.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    this.snapshot = snapshot;
+    this.logo = logo;
+    this.logoSmall = logoSmall;
+    this.textView = tv;
+    this.textViewShort = tvShort;
+    this.margin = margin;
+  }
+
+  public AttributionLayout measure() {
+    float logoContainerWidth = logo.getWidth() + (2 * margin);
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/attribution/AttributionMeasure.java]: `measure()` is kinda long. What about extracting each block of code into a private method?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/attribution/AttributionMeasure.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    float maxSizeShort = snapshot.getWidth();
+    boolean fullLogoText = logoContainerWidth + textViewContainerWidth <= maxSize;
+    boolean fullLogoShortText = logoContainerWidth + textViewShortContainerWidth <= maxSizeShort;
+    boolean smallLogoText = smallLogoContainerWidth + textViewContainerWidth <= maxSize;
+    boolean smallLogoShortText = smallLogoContainerWidth + textViewShortContainerWidth <= maxSizeShort;
+    boolean text = textViewContainerWidth + margin <= maxSize;
+    boolean shortText = textViewShortContainerWidth + margin <= maxSizeShort;
+    shorterText = fullLogoShortText || smallLogoShortText || shortText;
+
+    if (fullLogoText) {
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/attribution/AttributionMeasure.java]: What about removing the `if else if` chain using a map/dictionary?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/attribution/AttributionParser.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+  }
+
+  /**
+   * Invoked to validate if an url is valid for the improve map configuration.
+   *
+   * @param url the url to be validated
+   * @return if the url is valid for improve this map
+   */
+  private boolean isValidForImproveThisMap(String url) {
+    return withImproveMap || !url.equals("https://www.mapbox.com/map-feedback/");
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/attribution/AttributionParser.java]: Magic number

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/attribution/AttributionParser.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+  }
+
+  /**
+   * Invoked to validate if an url is valid for the Mapbox configuration.
+   *
+   * @param url the url to be validated
+   * @return if the url is valid for Mapbox
+   */
+  private boolean isValidForMapbox(String url) {
+    return withMapboxAttribution || !url.equals("https://www.mapbox.com/about/maps/");
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/attribution/AttributionParser.java]: Magic number

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/attribution/AttributionParser.java [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    }
+    return anchor;
+  }
+
+  /**
+   * Invoked to manually add attributions
+   */
+  private void addAdditionalAttributions() {
+    if (withTelemetryAttribution) {
+      String telemetryKey = "Telemetry Settings";
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/attribution/AttributionParser.java]: Magic numbers

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/snapshotter/MapSnapshotter.java [hunk 6]
```diff
@@ -269,43 +279,114 @@ public void cancel() {
...
+      .setLogo(logo.getLarge())
+      .setLogoSmall(logo.getSmall())
+      .setTextView(longText)
+      .setTextViewShort(shortText)
+      .setMarginPadding(margin)
+      .build();
+
+    AttributionLayout layout = measure.measure();
+
+    // draw logo
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/snapshotter/MapSnapshotter.java]: `addOverlay` is getting long.
What about extracting these blocks of code into private methods and give them a name based on the comments? This way `addOverlay` will be easier to read and understand and comments will become superfluous so they won't be necessary.

File: platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/maps/attribution/AttributionLayoutTest.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+      ViewGroup.LayoutParams.WRAP_CONTENT,
+      ViewGroup.LayoutParams.WRAP_CONTENT)
+    );
+    textView.setSingleLine(true);
+    textView.setTextSize(8);
+    textView.setText(ATTRIBUTION);
+  }
+
+  @Test
+  @Ignore
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/maps/attribution/AttributionLayoutTest.java]: All the tests are ignored, could we remove `AttributionLayoutTest`?

File: platform/android/MapboxGLAndroidSDK/src/test/java/com/mapbox/mapboxsdk/attribution/AttributionParseTest.java [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+      .build();
+
+    Set<Attribution> attributionList = attributionParser.getAttributions();
+    assertEquals("Size of list should match", 4, attributionList.size());
+
+    int counter = 0;
+    for (Attribution attribution : attributionList) {
+      switch (counter) {
+        case 0:
+          assertEquals("URL mapbox should match", "https://www.mapbox.com/about/maps/", attribution.getUrl());
```
Comment (Guardiola31337) [inline platform/android/MapboxGLAndroidSDK/src/test/java/com/mapbox/mapboxsdk/attribution/AttributionParseTest.java]: What about extracting the different assertions into different tests and give them a name based on the messages?
This way I guess that the `for` loop won't be necessary and tests will be more maintainable and readable.

Comment (tobrun) [reply platform/android/MapboxGLAndroidSDK/src/test/java/com/mapbox/mapboxsdk/attribution/AttributionParseTest.java]: don't have bandwith to pick this up. feel free to ticket out any follow up work.

**Pull request merged**